### PR TITLE
(onpremise) Fix `TypeError: a float is required` in 0024_auto_20191230_2052.py

### LINKED ISSUE #366: Upgrade 9.1.2 to 10. Migration error
Hello!
I've started to use sentry from version 9.0.0. On prod running sentry 9.1.2 now.
I decided install stage with onpremise. Also migrated database from prod. Version 9.1.2 works fine.
When upgrading to 10 I've got an error during migration 0024_auto_20191230_2052:
```
Operations to perform:
  Apply all migrations: admin, auth, contenttypes, jira_ac, nodestore, sentry, sessions, sites, social_auth
Running migrations:
  Applying sentry.0024_auto_20191230_2052.../usr/local/lib/python2.7/site-packages/django/db/models/fields/__init__.py:1451: RuntimeWarning: DateTimeField Event.datetime received a naive datetime (2019-10-29 11:37:43.582226) while time zone support is active.
  RuntimeWarning)
/usr/local/lib/python2.7/site-packages/django/db/models/fields/__init__.py:1451: RuntimeWarning: DateTimeField Event.datetime received a naive datetime (2020-01-27 11:37:43.582226) while time zone support is active.
  RuntimeWarning)
Events to process: 131482

Traceback (most recent call last):
  File "/usr/local/bin/sentry", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python2.7/site-packages/sentry/runner/__init__.py", line 164, in main
    cli(prog_name=get_prog(), obj={}, max_content_width=100)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/sentry/runner/decorators.py", line 30, in inner
    return ctx.invoke(f, *args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/sentry/runner/commands/upgrade.py", line 168, in upgrade
    _upgrade(not noinput, traceback, verbosity, not no_repair)
  File "/usr/local/lib/python2.7/site-packages/sentry/runner/commands/upgrade.py", line 130, in _upgrade
    ignore_ghost_migrations=True,
  File "/usr/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 131, in call_command
    return command.execute(*args, **defaults)
  File "/usr/local/lib/python2.7/site-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python2.7/site-packages/django/core/management/commands/migrate.py", line 204, in handle
    fake_initial=fake_initial,
  File "/usr/local/lib/python2.7/site-packages/django/db/migrations/executor.py", line 115, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/usr/local/lib/python2.7/site-packages/django/db/migrations/executor.py", line 145, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/usr/local/lib/python2.7/site-packages/sentry/new_migrations/monkey/executor.py", line 26, in apply_migration
    state, migration, fake=fake, fake_initial=fake_initial
  File "/usr/local/lib/python2.7/site-packages/django/db/migrations/executor.py", line 244, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/usr/local/lib/python2.7/site-packages/django/db/migrations/migration.py", line 129, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/usr/local/lib/python2.7/site-packages/django/db/migrations/operations/special.py", line 193, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/usr/local/lib/python2.7/site-packages/sentry/migrations/0024_auto_20191230_2052.py", line 70, in backfill_eventstream
    primary_hash = event.get_primary_hash()
AttributeError: 'Event' object has no attribute 'get_primary_hash'
```

How can I fix it?

### PR DESCRIPTION
**Pull request by ThomasWunderlich**
Fixes: https://github.com/getsentry/onpremise/issues/366

**Description of bug**
If an on-premise user has migrated rapidly (within the past 90 days) from 9.0.0 to 9.1.2 to 10.x they may encounter a `TypeError: a float is required` error because the event.timestamp does not exist and None is not a valid posix timestamp. 

The fix I've implemented wraps event.datetime in a try except block, and if it gets a type error uses the recieved_timestamp instead

**Alternative approaches considered:**
Something I considered, which may be cleaner from a maintainability perspective is overriding the insert logic in the migration, that way you don't have this try except block in a widely used code path

### CODE REVIEW

**Comment by ThomasWunderlich** [comment]
@BYK, @lynnagara this fixes a different edge case in migration 0024_auto_20191230_2052.py than https://github.com/getsentry/sentry/pull/16770 did. Please let me know if there are sql queries I can run to get you some sample data for testing. I've tested this with a database backup and a new server using master from yesterday


BTW, you may want to consider adding a PULL_REQUEST_TEMPLATE.md to .github. You already have a template for Issues.

**Comment by lynnagara** [comment]
Hi @ThomasWunderlich, thanks for the contribution.

Please edit only that particular migration (0024) instead of the eventstream implementation, which may have other consequences. The eventstream should not be responsible for dealing with historical event states. You can achieve this by ensuring that the `event` object is valid and has all its fields correctly populated in that migration before passing it to `eventstream.insert()`. Also the event `timestamp` is not the same as it's `received` (one is when it originally occurred and one when it got received by Sentry). It would be better to get the `timestamp` value from corresponding `datetime` column in the postgres table to preserve the original values rather than just making it the same as received.

File: src/sentry/migrations/0024_auto_20191230_2052.py
Comment (BYK) [inline src/sentry/migrations/0024_auto_20191230_2052.py:65]: FYI we are working towards a migration to Python 3.6 and beyond so I hope this works in them too?

File: src/sentry/migrations/0024_auto_20191230_2052.py
Comment (BYK) [inline src/sentry/migrations/0024_auto_20191230_2052.py:67]: Don't think `e` is defined here?

File: src/sentry/migrations/0024_auto_20191230_2052.py
Comment (BYK) [inline src/sentry/migrations/0024_auto_20191230_2052.py:84]: Any reason for not bundling this into `_attach_related`?

**Review by BYK — COMMENTED** [review]
Deferring to @lynnagara

File: src/sentry/migrations/0024_auto_20191230_2052.py
Comment (ThomasWunderlich) [reply src/sentry/migrations/0024_auto_20191230_2052.py:59]: Yes, this works going forward as well. The main consideration is that python 3.3+ has a function that converts a datetime to a timestamp without doing the ugliness below.

File: src/sentry/migrations/0024_auto_20191230_2052.py
Comment (ThomasWunderlich) [reply src/sentry/migrations/0024_auto_20191230_2052.py]: I've updated this per your suggestion so we avoid yet another loop through the events.

File: src/sentry/migrations/0024_auto_20191230_2052.py
Comment (ThomasWunderlich) [reply src/sentry/migrations/0024_auto_20191230_2052.py]: Ah you're right. I had forgotten to update this after I moved it into the callback

**Comment by ThomasWunderlich** [comment]
@lynnagara @BYK, This is ready to review. I've tested it with a database backup with old events and it hasn't hit any errors. Let me know if you want any further changes or want some sample test events .

File: src/sentry/migrations/0024_auto_20191230_2052.py
Comment (lynnagara) [reply src/sentry/migrations/0024_auto_20191230_2052.py:59]: I think you can leverage the `to_timestamp` helper here (and remove the python 2.7 comment) https://github.com/getsentry/sentry/blob/master/src/sentry/utils/dates.py#L15-L23

**Review by lynnagara — COMMENTED** [review]
I haven't run this code, but it makes sense to me. Just one suggestion about converting the timestamp

File: src/sentry/migrations/0024_auto_20191230_2052.py
Comment (ThomasWunderlich) [reply src/sentry/migrations/0024_auto_20191230_2052.py:59]: Ah perfect! I'll update now and then run through a test cycle

File: src/sentry/migrations/0024_auto_20191230_2052.py
Comment (lynnagara) [inline src/sentry/migrations/0024_auto_20191230_2052.py:59]: nit:`"timestamp" not in event.data.data`

**Review by lynnagara — APPROVED** [review]
Thanks @ThomasWunderlich!

**Comment by ThomasWunderlich** [comment]
I'll merge this in once I've finished testing. Thanks for all the time and help!

**Comment by lynnagara** [comment]
fyi - the failing travis test is unrelated and is being looked into, may need to wait for that to be fixed before merging

**Comment by ThomasWunderlich** [comment]
Happy to report that running this on our db backup was successful. Feel free to merge this in once the issue with the travis test is found

**Pull request merged**
